### ISSUE TITLE
AutoType keyboard layout

### ISSUE THREAD

**Issue by fneu** [firecat53/keepmenu#2]
Hey, I'm using a German keyboard layout (`de-latin1-nodeadkeys`) and keepmenu is auto-typing my usernames and passwords wrong.

An email address like `user@domain.com` is entered as `user"domain.com`. The `"` on a German keyboard is typed as `[shift]+[2]`, that's the location where `@` is on an american layout IIRC.

I didn't find a related setting in the config.ini, is it possible to use keepmenu with a German keyboard layout?

**Comment by firecat53** [comment]
Hmm, that would be a property of PyuserInput. It's unclear how support for alternate layouts after some initial research. There's another project called pynput...I'll put together a branch using that library instead and have you test it and see if it works any better/different.

**Comment by fneu** [comment]
Hey this looks like it could work:

PyUserInput:
```
>>> from pykeyboard import PyKeyboard
>>> k = PyKeyboard()
>>> k.type_string('@yzüöä{[]}')
"yzüöä(89)>>> 
```

Pynput
```
>>> from pynput.keyboard import Key, Controller
>>> keyboard = Controller()
>>> keyboard.type('@yzüöä{[]}')
@yzüöä{[]}>>> 
```

**Comment by fneu** [comment]
But even Pynput cannot type some characters, like the € sign...
meanwhile PyUserInput completely craps out on that one.

```
>>> from pykeyboard import PyKeyboard
>>> k = PyKeyboard()
>>> k.type_string('But can it type €?')
But can it type Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/fabian/.venv/input/lib/python3.6/site-packages/pykeyboard/base.py", line 79, in type_string
    self.tap_key(char)
  File "/home/fabian/.venv/input/lib/python3.6/site-packages/pykeyboard/base.py", line 43, in tap_key
    self.press_key(character)
  File "/home/fabian/.venv/input/lib/python3.6/site-packages/pykeyboard/x11.py", line 73, in press_key
    self._handle_key(character, X.KeyPress)
  File "/home/fabian/.venv/input/lib/python3.6/site-packages/pykeyboard/x11.py", line 66, in _handle_key
    fake_input(d, event, keycode)
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/home/fabian/.venv/input/lib/python3.6/site-packages/pymouse/x11.py", line 58, in manager
    raise X11Error(errors)
pymouse.x11.X11Error: [(BadValue(<Xlib.display._BaseDisplay object at 0x7fc257c3d1d0>, b'\x00\x02Q\x00\x00\x00\x00\x00\x02\x00\x84\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'), None)]
>>> 
```

```
>>> from pykeyboard import PyKeyboard
>>> from pynput.keyboard import Key, Controller
>>> keyboard = Controller()
>>> keyboard.type('But can it type €?') 
But can it type ?>>> 
```
**The crash of PyUserInput seems to also affect keepmenu**. After selecting an entry that contains an € sign, the entry is being typed up to the character before that symbol and then stops. Keepmenu then doesn't respond to further cli calls to it until it is killed and restarted. You might want to check out if that also happens for american keyboard layouts.

---

If there is no method to reliably type in special characters from python, you could either add copying to clipboard as an option, or maybe perform the typing via an external call to xdotool? It seems to handle the above cases fine.

**Comment by firecat53** [comment]
Thanks for the troubleshooting! I'm at work today, so unable to do much...

Great idea about calling xdotool as an alternative... I'll take a look at that. 

If you're ok with it, I'd suggest submitting your findings as bugs for both the py input projects and link them to this issue.

**Comment by firecat53** [comment]
Give the develop branch a try...I added xdotool support. Add an entry `type_library = xdotool` in the `database` section of config.ini. Let me know how it works!

**Comment by Gelion01** [comment]
Hey,
I followed this conversation, having the same issue with qwertz layout. I cloned your develop branch. xdotool usually makes the same mistake with irgnoring the keyboard layout. 

I solved this issue with calling "setxkbmap de" within the type_entry_xdotool function, this seems to be the common workaround to get xdotool to type things properly with qwertz (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=740811).

I am not that into programming, so let me know if this workaround isn't a proper option.

**Comment by firecat53** [comment]
Thanks for testing...lemme add that and have you test again and see if it works for you. Might be a day or two.

**Comment by firecat53** [comment]
@Gelion01  and @fneu, I've been playing with this and I think if you actually call 'setxkbmap xx' in your login profile someplace that it should carry through and be used by xdotool because they share the same X session.

My issue with calling setxkbmap from keepmenu is that it overrides any existing keyboard setups and Xmodmap mappings. This would definitely be an issue for people with more complex keyboard setups.

Can you both please try calling setxkbmap from your .bashrc, .bash_profile, .xinitrc, or whatever other file is called when you login (depending on your window manager/desktop environment) and then see if that is propagated through to xdotool called from keepmenu?

Thanks!

**Comment by Angerstoner** [comment]
Hi, first of all: thanks for your great work!

I'm using the develop branch and I'm executing "setxkbmap de" in my i3 config on i3 startup. Typing with xdotool works for me.

Keep it up and big thanks!

**Comment by firecat53** [comment]
Awesome, thanks for the comment! I'm gonna wait for a bit longer and see if anyone else can confirm this too before I merge the changes into master.

**Comment by Gelion01** [comment]
Same here. Also using i3 and not experiencing any problems with setting xkbmap via i3conf.

Thank you for your work!

**Comment by firecat53** [comment]
I went ahead and merged xdotool support into master and released 0.5.2 on PyPi. @fneu, let me know if there are any additional issues. Closing this for now. Thanks all!
